### PR TITLE
First pass at Storage conformance tests

### PR DESCRIPTION
**Pull request by jskeet**
The validation code is crude, but is good enough for bootstrapping.

### CODE REVIEW

File: storage/v1/test.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
+syntax = "proto3";
```
Comment (frankyn) [inline storage/v1/test.proto]: Copyright is needed here.

File: storage/v1/validation/Validation.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (frankyn) [inline storage/v1/validation/Validation.csproj]: Copyright is needed here.

File: storage/v1/test.proto (hunk 1 above)
Comment (jskeet) [reply storage/v1/test.proto]: Done.

File: storage/v1/validation/Validation.csproj (hunk 2 above)
Comment (jskeet) [reply storage/v1/validation/Validation.csproj]: This is just a project file - we've normally not added copyright for those. (Just like we don't for pom.xml in Java, for example.)

Comment (frankyn) [reply storage/v1/validation/Validation.csproj]: Interesting, for Java samples (java-docs-samples), the copyright is added to the pom.xml files, but is not for google-cloud-java. 

C# samples don't add it to the csproj file as well. Thanks, please ignore this comment.

**Comment by frankyn** [comment]
@jskeet, PR lgtm. However, I'd like to wait until the repo has tests with Kokoro CI working to merge. I'll aim for this week, What do you think? 

Thanks for your patience!

**Comment by jskeet** [comment]
I've got other things I can be doing, certainly - it won't block me.

**Comment by frankyn** [comment]
Sg, thank you!

**Comment by jskeet** [comment]
Whoops! Yup, will do so in the morning.

**Comment by jskeet** [comment]
Added in another commit.

**Comment by jadekler** [comment]
As with others on the team (justin and co), I agree that it'd be nice to use kokoro instead of travis. We've heavily prioritized using kokoro over travis as a team and it seems a shame to go back on that just for the sake of .NET not being supported yet / ~30 lines of code.

Here's the same validation program in Go, if we feel like temporarily swapping languages in order to forego travis:

```
package main

import (
	"bytes"
	"io/ioutil"
	"strings"

	"github.com/golang/protobuf/jsonpb"
	storage_v1_tests "/path/to/generate/.pb.go"
)

func main() {
	dir := "internal/conformance"

	files, err := ioutil.ReadDir(dir)
	if err != nil {
		panic(err)
	}

	for _, f := range files {
		if !strings.Contains(f.Name(), ".json") {
			continue
		}

		inBytes, err := ioutil.ReadFile(dir + "/" + f.Name())
		if err != nil {
			panic(err)
		}

		var testfile storage_v1_tests.TestFile
		if err := jsonpb.Unmarshal(bytes.NewBuffer(inBytes), &testfile); err != nil {
			panic(err)
		}
	}
}
```

**Comment by jskeet** [comment]
Fine by me. My only reason for using C# was to make it simpler to throw together a prototype.

**Comment by frankyn** [comment]
Thanks @jskeet, I'll move forward on the Go version and close this PR. 

Thanks @jadekler!

To answer your questions, I've responded inline below.

> Some things that seem to be wrong in ours:
> 
>         want:	https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=host&X-Goog-Signature=95e6a13d43a1d1962e667f17397f2b80ac9bdd1669210d5e08e0135df9dff4e56113485dbe429ca2266487b9d1796ebdee2d7cf682a6ef3bb9fbb4c351686fba90d7b621cf1c4eb1fdf126460dd25fa0837dfdde0a9fd98662ce60844c458448fb2b352c203d9969cb74efa4bdb742287744a4f2308afa4af0e0773f55e32e92973619249214b97283b2daa14195244444e33f938138d1e5f561088ce8011f4986dda33a556412594db7c12fc40e1ff3f1bedeb7a42f5bcda0b9567f17f65855f65071fabb88ea12371877f3f77f10e1466fff6ff6973b74a933322ff0949ce357e20abe96c3dd5cfab42c9c83e740a4d32b9e11e146f0eb3404d2e975896f74
>         got:	https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190424%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190424T210206Z&X-Goog-Expires=-1556139716&X-Goog-Signature=4a3f46c3468284778b3f2d175656ab3693bf4f1a49577a4b76734e555f4abe9b20f3b8a3ebf167f323149d4ed006434463f7c0c6d4fbed1c1986ccd27da000383bb3f0265ad7d68825a229d27578bde9b69d37ef32e4e63a0f55149b64cceab308253618d80c6e210a334f6c425493f0323200ba5fefe897e51331eb770ee2b450cb721c7f39b95b537d9a9aa1614dbe39f3787482a175cc5129ac1f9f63422c2fb6cfa5ab6d3cf7e51d82de2ef2dad10c49bd8d072b5521ac4a69b988626750ab3ee19a30ba27450a23831584ebb2d491112a14f0960741d58d16f1e81e908935d414cdce61d231d684fb3bfd963aff5531da2606553fae1db4c7530e1ce113&X-Goog-SignedHeaders=host

Issues with "got"
`X-Goog-SignedHeaders` should only show up once.
`X-Goog-Expires` expects seconds from X-Goog-Date.
`X-Goog-Signature` will be impacted if anything is different in the construction of the string_to_sign.

> Time is not able to be hardcoded. We're going to have to do string substitution. Can we use an obviously dummy time in timestamp and expectedUrl?

Could you elaborate what you mean by have to do a string substitution? The "timestamp" value in the conformance test is the value of X-Goog-Date and is the "dummy time".

> We appear to have X-Goog-SignedHeaders=host (twice, for some reason). Is that wrong?

`X-Goog-SignedHeaders` should only show up once.

> Our expiration seems to be calculated from time.Now() instead of just a static value. That is, it seems to say "expires in this time" rather than "expires at this time".

`time.Now()` will need to be updated to the dummy timestamp provided in the conformance test.

> Our X-Goog-Signature differs from yours. I naively assumed that you're supplying a private key with the value that's in the service account json. Is that right, or?

Yes, the private key in the service-account JSON file is the expected private key. Go client requires additional parsing in the default case.

**Comment by jskeet** [comment]
@frankyn: Did you mean to just close this PR? What's the plan?

**Comment by frankyn** [comment]
My plan for today is to create another PR to introduce the conformance tests into the repo based on @jadekler's Go version and include Kokoro CI for Go.

File: storage/v1/signing.json [hunk 3]
```diff
@@ -0,0 +1,113 @@
+{
+  "signingV4Tests": [
+    {
+      "description": "Simple GET",
+      "bucket": "test-bucket",
+      "object": "test-object",
+      "method": "GET",
+      "expiration": "10",
+      "timestamp": "2019-02-01T09:00:00Z",
+      "expectedUrl": "https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256\u0026X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request\u0026X-Goog-Date=20190201T090000Z\u0026X-Goog-Expires=10\u0026X-Goog-SignedHeaders=host\u0026X-Goog-Signature=95e6a13d43a1d1962e667f17397f2b80ac9bdd1669210d5e08e0135df9dff4e56113485dbe429ca2266487b9d1796ebdee2d7cf682a6ef3bb9fbb4c351686fba90d7b621cf1c4eb1fdf126460dd25fa0837dfdde0a9fd98662ce60844c458448fb2b352c203d9969cb74efa4bdb742287744a4f2308afa4af0e0773f55e32e92973619249214b97283b2daa14195244444e33f938138d1e5f561088ce8011f4986dda33a556412594db7c12fc40e1ff3f1bedeb7a42f5bcda0b9567f17f65855f65071fabb88ea12371877f3f77f10e1466fff6ff6973b74a933322ff0949ce357e20abe96c3dd5cfab42c9c83e740a4d32b9e11e146f0eb3404d2e975896f74"
```
Comment (jadekler) [inline storage/v1/signing.json]: @jskeet @frankyn Go's headers get ordered lexicographically. Can we put X-Goog-Signature before X-Goog-SignedHeaders?

Comment (frankyn) [reply storage/v1/signing.json]: @jadekler, other languages have conformed to appending the signature to the end.

Given the Go interface returns a string type, could you instead append X-Goog-Signature to the end of the resulting URL?

**Comment by jadekler** [comment]
Screen Shot 2019-04-26 at 8 31 56 AM

Cool so apparently I can't reply to conversation threads. Hooray github! :)

@frankyn Go builds a URL in a url.Url and then calls url.String on it, basically. We don't construct the URL manually. We could parse and reformat the URL in the tests rather than doing string matching on it, but it seems clunky... if all other 6 languages prefer not to change their impl maybe that's the best approach though. WDYT?

**Comment by jskeet** [comment]
I don't mind moving things around for C#.

**Comment by frankyn** [comment]
Hey @jadekler,

I'd like to add few more items of why we shouldn't update the format. I realized my prior reasoning was lacking.

- [Documentation](https://cloud.google.com/storage/docs/access-control/signed-urls) uses the format with `X-Goog-Signature` at the end of the URL.
- [gsutil](https://cloud.google.com/storage/docs/access-control/signing-urls-with-helpers) generates signed URLs in this format. 
- It's also a common industry format.

I'm hesitant of changing the format based on one language with respect to the information across the parts that make up this feature. Additionally, we can't change related documentation outside of our control. 

Apologies, I don't mean to create more friction with the Go implementation.

**Pull request closed without merging**